conclave-enclave/src/main/kotlin/com/r3/conclave/enclave/internal/GramineSGXEnclaveEnvironment.kt

```diff
@@ -0,0 +1,119 @@
+        reportData: ByteCursor<SgxReportData>?
+    ): ByteCursor<SgxReport> {
+        val report = retrieveReport(
+            targetInfo?.buffer?.getRemainingBytes(avoidCopying = true),
+            reportData?.buffer?.getRemainingBytes(avoidCopying = true)
+        )
+        return Cursor.slice(SgxReport, ByteBuffer.wrap(report))
+    }
+
+    override fun getSignedQuote(
+        quotingEnclaveInfo: ByteCursor<SgxTargetInfo>?,
+        reportData: ByteCursor<SgxReportData>?
+    ): ByteCursor<SgxSignedQuote> {
+        //  Note that in Gramine the "signed quote" is automatically retrieved and returned
+        //    by the enclave. There is no enclave to host communication that we need to handle in our code.
+        //    In the background, Gramine interacts with the AESM service and the quoting enclave
+        //    to get the "signed quote".
+        //  In Graal VM/Native Image flow the "quotingEnclaveInfo" (also called "SGX target info") is
+        //    requested by the host to the quoting enclave by passing an empty array to the
+        //    function "sgx_get_target_info", which fills that array.
+        //  When working with the Gramine flow, such operation is done on the enclave side by passing an empty
+        //    array that it is filled by Gramine (which communicates with the quoting enclave in background).
+        val quotingEnclaveInfoBytes = quotingEnclaveInfo?.bytes ?: Cursor.allocate(SgxTargetInfo).bytes
+        createReport(ByteCursor.wrap(SgxTargetInfo, quotingEnclaveInfoBytes), reportData)
+        val signedQuoteBytes = readSignedQuote()
+        return Cursor.slice(SgxSignedQuote, ByteBuffer.wrap(signedQuoteBytes))
+    }
+
+    override fun sealData(toBeSealed: PlaintextAndEnvelope): ByteArray {
+        //  TODO: Gramine filesystem support
+        return EnclaveUtils.sealData(aesSealingKey, toBeSealed)
+    }
+
+    override fun unsealData(sealedBlob: ByteBuffer): PlaintextAndEnvelope {
+        //  TODO: Gramine filesystem support
+        return EnclaveUtils.unsealData(aesSealingKey, sealedBlob)
+    }
+
+    override fun getSecretKey(keyRequest: ByteCursor<SgxKeyRequest>): ByteArray {
+        //  TODO: Gramine filesystem support
+        //  Note that we will probably not need to pass around the local secret key
+        //    as this one will be embedded in the manifest
+        //  https://gramine.readthedocs.io/en/stable/manifest-syntax.html#encrypted-files
+        return byteArrayOf()
+    }
+
+    override fun setupFileSystems(
+        inMemoryFsSize: Long,
+        persistentFsSize: Long,
+        inMemoryMountPath: String,
+        persistentMountPath: String,
+        encryptionKey: ByteArray
+    ) {
+        //  TODO: Gramine filesystem support
+    }
+
+    private fun retrieveReport(targetInfoBytes: ByteArray?, userReportDataBytes: ByteArray?): ByteArray {
+        if (targetInfoBytes != null) {
+            writeTargetInfo(targetInfoBytes)
+        }
```

Comment:
>user report data must be written into /dev/attestation/user_report_data and target info must be written into /dev/attestation/target_info. Otherwise the obtained report will contain incorrect or stale user report data and target info.

This is what I'm talking about. Even if the user doens't provide values for these, the resulting report must not use old values. It's insecure.

Reply:
This code will also suffer from race conditions if multiple threads try to get a report. Annotate this method with `@Synchronized`.